pkg/allocateip/allocateip.go

```diff
@@ -201,48 +266,50 @@ func assignHostTunnelAddr(ctx context.Context, c client.Interface, nodename stri
 		logCtx.Fatal("Unable to autoassign an address - pools are likely exhausted.")
 	}
 
-	var updateError error
+	// Update the node object with the assigned address.
+	ip := ipv4Addrs[0].IP.String()
+	if err = updateNodeWithAddress(ctx, c, nodename, ip, vxlan); err != nil {
+		// We hit an error, so release the IP address before exiting.
+		err := c.IPAM().ReleaseByHandle(ctx, handle)
+		if err != nil {
+			logCtx.WithError(err).WithField("IP", ip).Errorf("Error releasing IP address on failure")
+		}
+
+		// Log the error and exit with exit code 1.
+		logCtx.WithError(err).WithField("IP", ip).Fatal("Unable to set tunnel address")
+	}
+
+	logCtx.WithField("IP", ip).Info("Set tunnel address")
+}
+
+func updateNodeWithAddress(ctx context.Context, c client.Interface, nodename string, addr string, vxlan bool) error {
```

Comment:
GitHub is rendering this diff a bit strangely, but all I did was pull out the logic which updates the node into a helper function so that error handling was a bit more sane.